File: FirebaseFirst/app/src/main/java/com/frederico/firebasefirst/MainActivity.java
package com.frederico.firebasefirst;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.frederico.firebasefirst.adapters.TarefaAdapter;
import com.frederico.firebasefirst.models.TarefaModel;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.GenericTypeIndicator;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {
    private DatabaseReference reference = FirebaseDatabase.getInstance().getReference();
    Toolbar toolbar;
    ArrayList<TarefaModel> lista = new ArrayList<>();
    RecyclerView recyclerView;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        recyclerView = findViewById(R.id.recycler);

        fillAdapter();

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        //Log.i("QUERO MORRE", "A");
        Intent intent = new Intent(this, CriarEditar.class);
        startActivity(intent);
        return super.onOptionsItemSelected(item);
    }

    public void fillAdapter(){
        reference.child("tarefas").addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                String id,title, content;
                for(DataSnapshot snap: snapshot.getChildren()){
                    id = snap.getKey();
                    title = snap.child("titulo").getValue().toString();
                    content = snap.child("conteudo").getValue().toString();
                    lista.add(new TarefaModel(id, title, content, new pressionar()));
                }
                TarefaAdapter adapter = new TarefaAdapter(lista);
                LinearLayoutManager managero = new LinearLayoutManager(getApplicationContext());
                recyclerView.setHasFixedSize(true);
                recyclerView.setLayoutManager(managero);
                recyclerView.addItemDecoration(new DividerItemDecoration(getApplicationContext(), LinearLayout.VERTICAL));
                recyclerView.setAdapter(adapter);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }

    class pressionar implements TarefaModel.Press{

        @Override
        public void Short(String id) {
            Intent intent = new Intent(getApplicationContext(), CriarEditar.class);
            intent.putExtra("id", id);
            startActivity(intent);
        }

        @Override
        public void Long(String id) {
            AlertDialog.Builder alert = new AlertDialog.Builder(MainActivity.this);
            alert.setTitle("Atenção");
            alert.setIcon(R.drawable.ic_baseline_warning_24);
            alert.setMessage("Você tem certeza que deseja deletar esta Tarefa?");
            boolean response = false;
            alert.setPositiveButton("Sim", new DialogInterface.OnClickListener(){
                @Override
                public void onClick(DialogInterface dialogInterface, int i){
                    deleteTarefa(id);
                }
            });
            alert.setNegativeButton("Nao", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i) {

                }
            });
            alert.create().show();
        }
    }

    void deleteTarefa(String id){
        reference.child("tarefas").child(id).setValue(null);
        recreate();
    }
}

File: NotasKotlin/app/src/main/java/com/frederico/notaskotlin/MainActivity.kt
package com.frederico.notaskotlin

import android.annotation.SuppressLint
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.frederico.notaskotlin.adapters.NotaAdapter
import com.frederico.notaskotlin.model.Nota
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener

class MainActivity : AppCompatActivity() {
    val notas = mutableListOf<Nota>()
    lateinit var recyclerView: RecyclerView
    private val database = FirebaseDatabase.getInstance().reference

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        val adicionarNota = findViewById<ExtendedFloatingActionButton>(R.id.botao_nova_nota)

        getNotas()
        recyclerConfig(adicionarNota)

        adicionarNota.setOnClickListener {
            /*
            database.push().setValue(
                hashMapOf(
                    "titulo" to "titulo de dentro",
                    "texto" to "texto de dentro",
                    "label" to 0
                )
            )
            */
            val intent = Intent(this, EditNote::class.java) //.putExtra("ID", "TESTE")
            startActivity(intent)
        }

    }

    @SuppressLint("NotifyDataSetChanged")
    fun deletar(id: String){
        val nota = notas.find { it.id == id }
        database.child(nota!!.id).removeValue()
    }

    fun editar(id : String){
        val intent = Intent(this, EditNote::class.java)
        intent.putExtra("ID", id)
        startActivity(intent)
    }

    private fun getNotas(){
        database.addValueEventListener(object: ValueEventListener{
            @SuppressLint("NotifyDataSetChanged")
            override fun onDataChange(snapshot: DataSnapshot) {
                for(snap in snapshot.children){
                    println(snap)
                    if(snap.hasChild("titulo") and snap.hasChild("texto") and snap.hasChild("label") ) {
                        if (notas.find { it.id == snap.key.toString() } != null) {
                            notas.find { it.id == snap.key.toString() }?.titulo =
                                snap.child("titulo").value.toString()
                            notas.find { it.id == snap.key.toString() }?.texto =
                                snap.child("texto").value.toString()
                            notas.find { it.id == snap.key.toString() }?.label =
                                snap.child("label").value.toString().toInt()
                        } else {
                            notas.add(
                                Nota(
                                    snap.key.toString(),
                                    snap.child("texto").value.toString(),
                                    snap.child("titulo").value.toString(),
                                    ::deletar,
                                    ::editar,
                                    snap.child("label").value.toString().toInt()
                                )
                            )
                        }
                    }
                }

                val databaseids = snapshot.children.map { it.key.toString() }.toSet()
                notas.removeAll(notas.filter { it.id !in databaseids })
                recyclerView.adapter?.notifyDataSetChanged()
            }

            override fun onCancelled(error: DatabaseError) {
            }
        })
    }

    fun recyclerConfig(botao: ExtendedFloatingActionButton){
        recyclerView = findViewById(R.id.recycler_view)
        recyclerView.adapter = NotaAdapter(notas)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.addOnScrollListener( object: RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                println(dy)
                if(dy > 0){
                    botao.hide()
                }else if(dy < -9 || dy == 0){
                    botao.show()
                }
                super.onScrolled(recyclerView, dx, dy)
            }
        })
        recyclerView.setHasFixedSize(true)
    }
}